### ISSUE TITLE
Add configurable colorblind modes as a standard feature

### ISSUE THREAD

**Issue by alice-i-cecile** [bevyengine/bevy#2724]
## What problem does this solve or what need does it fill?

Color-blindness is common, and each game should have a solution to address this.

## What solution would you like?

We should address this by providing a common API for games to hook into. EA's newly released [Tunable Colorblindness Solution](https://github.com/electronicarts/Tunable-Colorblindness-Solution) looks like a promising base.

At first glance, this seems like it is probably most suitable for a fresh implementation as a post-processing shader to alter colors according to the specified parameters.

## What alternative(s) have you considered?

Use a different colorblindness approach.

Bake these changes into the asset pipeline?

## Additional context

The code is licensed under Apache 2.0. This shouldn't be an issue with a fresh implementation of the algorithms in Rust.

We should ensure that this is stored in a standalone crate if at all possible, allowing the rest of the ecosystem to reuse our work.

**Comment by Vrixyz** [comment]
I implemented the code you linked in bevy as a WIP example : https://github.com/Vrixyz/bevy/tree/colorblind.

I think we'll want to incorporate it behind a plugin but currently it's a bit complicated to do, I'll wait for feedbacks on https://github.com/bevyengine/bevy/pull/4797 before working on a colorblind plugin.

**Comment by alice-i-cecile** [comment]
Yep agreed, I'd like this as a rendering plugin, and then have the example just be a demo of how to set up a menu to enable this in your game.

**Comment by Elabajaba** [comment]
I'm very strongly against the inclusion of any sort of fullscreen tinting as a form of colourblind mode, or to even consider fullscreen tinting a form of colourblindness accessibility feature because it sucks. It generally just makes the game look ugly, it's already built into the OS, and I don't know of any colourblind people that like or use them becaues of how bad fullscreen tinting is as an accessibility feature.

What really needs to be done is to allow users to modify the colours and opacity of UI elements.

For an FPS this would mean allowing users to change the colours of the crosshair (including when it's not on anything, when it's on an enemy, when it's on an ally, etc) and changing the colour of the inworld UI markers (eg. the colour+opacity of objective markers, what colour your teammates+enemies healthbars/markers/etc are, if you have an ingame ping sytem the colour of the ping markers). Recent Battlefields have attempted this by allowing you to select different presets for the ingame markers (see https://i.imgur.com/wrALSPZ.jpg), but in my experience the colours they chose for the colourblind presets, while easier to differentiate between, blended into the game too much and ended up being harder to see (easily fixed by allowing users to choose their own colours). Dirty Bomb had a great system for this, despite not exposing any of it to the user through the options menu. They stored the UI element colour+opacity+position in config files, and allowed users to modify them (many players used this to create their own centered UI, which would also be useful for ultrawide users, see end of post for example).

For an RTS it would include allowing users to change what the team colours are, what the minimap colours are, and potentially what colour pings/flags are (eg. in Age of Empires 2 green and yellow are almost identical, orange blends into the minimap, and blue is impossible to see on the minimap on water maps. iirc the Definitive Edition included colour blind settings that changes the team colours, but the colours they change them to still have issues due to choosing colours that blend together on the minimap. Allowing users to change them easily to whatever colours the user wanted is the proper solution to this.)

For a MOBA, you need to allow users to change the colours of particle effects. When League of Legends was first released, they didn't have a colourblind toggle, so some players couldn't see eg. Singed's poison trail, meaning he was basically impossible to play against for them. When League added a colourblind toggle it was great. It solved a lot of issues people had with a lot of champions where they just couldn't see their abillities. However, over the past decade League has added a ton of skins that don't bother with a colourblind variation despite many players complaining about not being able to see the spell particles on a lot of skins. They've also added a map with a blue background instead of a green background, so the light blue tint they generally use to make spells more visible on Summoner's Rift (the default green/brown terrain map) makes spells blend in on Howling Abyss (blue/white snowy ARAM map).

<details>
<summary>Dirty bomb centered UI customization example</summary>

 https://i.imgur.com/pEa04T1.png

```
MinimapPosition=(Align=(H=HA_CENTER, V=HA_BOTTOM), Attach=(H=HA_CENTER, V=HA_BOTTOM), PixelOffset=(X=0, Y=65))
WeaponAmmoPosition=(Align=(H=HA_RIGHT, V=HA_BOTTOM), Attach=(H=HA_RIGHT, V=HA_TOP, VA="Minimap", HA="Minimap"), PixelOffset=(X=5, Y=0))
PlayerHealthPosition=(Align=(H=HA_RIGHT, V=HA_BOTTOM), Attach=(H=HA_RIGHT, V=HA_TOP, VA="WeaponAmmo", HA="WeaponAmmo"), PixelOffset=(X=-5, Y=40))
AbilityCooldownPosition=(Align=(H=HA_RIGHT, V=HA_TOP), Attach=(H=HA_LEFT, V=HA_TOP, VA="Minimap", HA="Minimap"), PixelOffset=(X=-5, Y=7))
GameWaveTimerPosition=(Attach=(H=HA_CENTER, V=HA_TOP), Align=(H=HA_CENTER, V=HA_BOTTOM, VA="Minimap", HA="Minimap"), PixelOffset=(X=133, Y=802))
ExpCounterPosition=(Align=(V=HA_TOP), Attach=(H=HA_CENTER, V=HA_TOP), PixelOffset=(Y=150))
```
</details>

**Comment by Vrixyz** [comment]
Thanks for the detailed feedback, as a non-colorblind these informations are very valuable to me.

I was not considering this issue to make as an “accessibility feature” for end users though, but a helper for developers, to get a feeling of what colorblind people would see when looking at the fullscreen, hence the fullscreen tinting.

A on/off tinting to different colorblindness conditions could really help realizing issues with the game, and adapt the game colors afterwards. That color fix-up would be out of scope of that fullscreen tinting and would probably have to be made manually and on a case-by-casy basis.

**Comment by bjorn3** [comment]
I think it should be called something like "emulate colorblindness" then.

**Comment by annieversary** [comment]
I recently published a crate that solves the "emulating colorblindness" part and I was asked to drop the link here: [bevy_color_blindness](https://crates.io/crates/bevy_color_blindness). It's based on [this](https://www.alanzucconi.com/2015/12/16/color-blindness/) article by Alan Zucconi.

I don't know if I have much else to contribute to the conversation as I am not color blind, but I fear that adding built-in colorblind modes will make developers assume they can check the Accessibility box on their TODO list by simply adding an option to enable it, when there's more to accessibility than that. I obviously don't want to say no to something that improves accessibility, but if it does get added into Bevy, it should have of documentation explaining how it's not a one stop solution, and that other options should be considered in addition to it.

**Comment by alice-i-cecile** [comment]
@annieversary that crate looks incredible. Would you be alright if we began the upstreaming process? I think it's well-contained and well-made enough that we can just incorporate it in a single PR.

**Comment by annieversary** [comment]
Sure! How does the process work? Do I make a PR that adds the whole crate inside `crates/`? Or would we want a `bevy_accessibility` crate that has it as a module?

**Comment by alice-i-cecile** [comment]
I'd like to avoid segmenting accessibility features into their own niche; IMO they should be well-integrated with the crates they're building on :)

In this case, it feels like this should be part of `bevy_render`, controlled by a feature flag that's enabled by default. And then we should have a top-level example in `bevy` demonstrating the feature :)

@superdump, please feel free to overrule me either here or in the PR though.

**Comment by annieversary** [comment]
That makes sense! 

I don't think it can be in `bevy_render`, since it depends on `Material2d` from `bevy_sprite` and `UiCameraConfig` from `bevy_ui`, and when I try to use them, Cargo complains about cyclic dependencies. It could be in `bevy_ui`, but I don't think it fits well conceptually.

**Comment by alice-i-cecile** [comment]
Alright :) Make it its own top-level crate for now, and we can refactor it later!

**Comment by annieversary** [comment]
Alright! I've opened the PR ^^
